docs/blueprints/EP038.rst

```diff
@@ -0,0 +1,71 @@
+:EP: 38
+:Title: Improving security to coloring application to switches
+:Authors:
+    - Aldo Ortega <aortegay AT fiu DOT edu>
+    - Italo Valcy <idasilva AT fiu DOT edu>
+
+:Created: 2023-06-27
+:Updated: 2023-06-27
+:Kytos-Version: 2023.2
+:Status: Draft
+
+*************************************************************
+EP38 - Improving security to coloring application to switches
+*************************************************************
+
+Abstract
+========
+
+Here it will be explained the vulnerability of coloring implementation along with some of the issues that these problems could have in a deployed topology. Also, it is going to be explained how these vulnerabilities could be removed.
+
+Motivation
+==========
+
+The NApp **coloring** installs flows according to a switch neighbors (switches connected by a link). These flows have three crucial characteristics:
+
+  1. The matching criteria has one field, **"dl_src"** which has the same standard as a **MAC** address; ie. **ee:ee:ee:ee:ee:02**.
+  2. The received packet is send to the controller.
+  3. They have the highest priority.
+
+A customer connection with a certain **MAC** address could send packets indefinitely overloading the controller and interrupting other flows in the switch. Finally the currently used switch could disconnect.
+
+Solution
+========
+
+We need to ensure that the packets that **coloring** is interested in, come from inside the topology and drop anything else. Adding the following should suffice:
+
+  - Adding **"in_port"** as matching criteria to the created flow. The proper port could be obtained from the connection of switches neighbors.
+  - Add an extra flow without **"in_port"** so it matches what is left. The matched packets need to be dropped to avoid further inturreptions with other flows.
```

Comment:
Do we really need this extra flow with `dl_src` and `actions=drop`? Based on OF1.3 spec:

> If the table-miss flow entry does not exist, by default packets unmatched by flow entries are dropped
(discarded). A switch configuration, for example using the OpenFlow Configuration Protocol, may override
this default and specify another behaviour.

Since your proposal is narrowing down the match to include `in_port` then, a matched `dl_src` from a `in_port` that doesn't have a flow rule won't match. Unless network engineers install a flow table miss, but that's outside of `coloring` responsibilities, so looks like we could get away without having to maintain this extra drop flow, right? Unless the intention is to also try to optimize the lookup when it gets dropped, but usually switches have internal optimization for their lookups, so not sure if it's worth the effort to maintain and measure this efficiency. Food for thought.

Reply:
Another related question: How will it detect if it's coming from "inside the topology"? Currently, `coloring` is trusting `of_lldp` so any discovered link will be consider to reach to an adjacent switch, and consequently that interface will have `nni: True`. 

So, maybe we should make it explicit that `of_lldp` shouldn't be enabled on interfaces where it doesn't expect to install `coloring` related flows? That would need confirmation if it'd be acceptable, if we need a higher level of control then maybe we might need a new API on `coloring` to have more control on which interfaces should be enabled.

Reply:
> How will it detect if it's coming from "inside the topology"?

Oh I meant that we could use the information that `coloring` has about link when creating flows. Link dictionaries contain the endpoints and the ports they are connected to. The idea is to use these ports for the matching field `in_port`.

Reply:
I see. At is is though, If this customer port is connected to a switch, and if it sends LLDP packets compatible with the flow entry that `of_lldp` installs then it would also end up as a link in our topology. Although if it's not a switch managed by kytos, then `kytos/of_lldp.interface.is.nni` wouldn't get discovered. 

It'd be great to clarify what we're meaning by customer connection in this case, is it also a switch managed by our controler or not? or can it be both? Depending on the scope we might need different guarantees.

Reply:
I see, will explain better on the blueprint. By customer I meant an external connection but I could not make it possible. Changing it to not NNI. So `coloring` will create flows only for discovered NNIs.

Reply:
I see. Right, if it `coloring` keeps going hand in hand with `of_lldp` and the discovered NNIs that facilitates.